### PR TITLE
Ser for msp podcast

### PR DESCRIPTION
**Pull request by Subohao**
## What?

Modify the template from slu1 and spk to ser downstream tasks, specifically for primary categorical emotions in msp-podcast, excluding X and O.

## Why?

For ESPnet BootCamp and integrating SER downstream tasks into ESPnet.

### CODE REVIEW

**Comment by sw005320** [comment]
Can you fix the CI error in https://github.com/espnet/espnet/actions/runs/14582314636/job/40901457851?pr=6096#step:8:285 ?

**Comment by sw005320** [comment]
I thought that you would implement it in the `slu` task.
Either is fine, but can you explain why you created a new task?

**Comment by Subohao** [comment]
> I thought that you would implement it in the `slu` task. Either is fine, but can you explain why you created a new task?

Based on my understanding of the SLU recipe, it's tied to the ASR task as well. In this SER recipe, I certainly simplify it into a simple emotion classification task by only considering the speech input and emotions as output, without tackling much on the speaker and text part of MSP-PODCAST. Also, in SER, we would usually have a pooling and a projection layer at the end of the model, which does not seem to be in the SLU recipe. Therefore, I created a new task to do it.

**Comment by sw005320** [comment]
@ftshijt, can you review this PR since it is a new task.

File: egs2/TEMPLATE/ser1/ser.sh [hunk 1]
```diff
@@ -0,0 +1,1677 @@
...
+                    --collect_stats true \
+                    --use_preprocessor true \
+                    --bpemodel "${bpemodel}" \
+                    --transcript_token_list "${transcript_token_list}" \
+                    --token_type "${token_type}" \
+                    --token_list "${token_list}" \
+                    --non_linguistic_symbols "${nlsyms_txt}" \
+                    --cleaner "${cleaner}" \
+                    --g2p "${g2p}" \
+                    --emotions "A_S_H_U_F_D_C_N_O_X" \
```
Comment (ftshijt) [inline egs2/TEMPLATE/ser1/ser.sh:1055]: Is emotion fixed only to these categories?

File: egs2/TEMPLATE/ser1/ser.sh [hunk 2]
```diff
@@ -0,0 +1,1677 @@
...
+        # shellcheck disable=SC2086
+        if "${use_transcript}"; then
+            ${train_cmd} JOB=1:"${_nj}" "${_logdir}"/stats.JOB.log \
+                ${python} -m espnet2.bin.ser_train \
+                    --collect_stats true \
+                    --use_preprocessor true \
+                    --bpemodel "${bpemodel}" \
+                    --transcript_token_list "${transcript_token_list}" \
+                    --token_type "${token_type}" \
+                    --token_list "${token_list}" \
```
Comment (ftshijt) [inline egs2/TEMPLATE/ser1/ser.sh:1051]: what is the difference between transcript token list and token list?

File: egs2/TEMPLATE/ser1/ser.sh [hunk 3]
```diff
@@ -0,0 +1,1677 @@
...
+fi
+
+
+# ========================== Data preparation is done here. ==========================
+
+
+if ! "${skip_train}"; then
+    if "${use_lm}"; then
+        if [ ${stage} -le 6 ] && [ ${stop_stage} -ge 6 ]; then
+            log "Stage 6: LM collect stats: train_set=${data_feats}/lm_train.txt, dev_set=${lm_dev_text}"
```
Comment (ftshijt) [inline egs2/TEMPLATE/ser1/ser.sh:787]: It seems that we do not need to have LM stages?

File: egs2/TEMPLATE/ser1/ser.sh [hunk 4]
```diff
@@ -0,0 +1,1677 @@
...
+                    --valid_shape_file "${_logdir}/valid.JOB.scp" \
+                    --output_dir "${_logdir}/stats.JOB" \
+                    ${_opts} ${ser_args} || { cat "${_logdir}"/stats.1.log; exit 1; }
+        else
+                    # --bpemodel "${bpemodel}" \
+                    # --token_type "${token_type}" \
+                    # --token_list "${token_list}" \
+                    # --non_linguistic_symbols "${nlsyms_txt}" \
+                    # --cleaner "${cleaner}" \
+                    # --g2p "${g2p}" \
```
Comment (ftshijt) [inline egs2/TEMPLATE/ser1/ser.sh:1071]: Please clean up the commented lines (also for following parts)

File: egs2/msp_podcast/ser1/cmd.sh [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+# If jobs failed, your configuration might be wrong for your environment.
+#
+#
+# The official documentation for run.pl, queue.pl, slurm.pl, and ssh.pl:
+#   "Parallelization in Kaldi": http://kaldi-asr.org/doc/queue.html
+# =========================================================~
+
+
+# Select the backend used by run.sh from "local", "stdout", "sge", "slurm", or "ssh"
+cmd_backend='slurm' #cmd_backend='local'
```
Comment (ftshijt) [inline egs2/msp_podcast/ser1/cmd.sh:31]: You would need to change it to local to pass the ci

File: egs2/msp_podcast/ser1/local/score_ser.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import argparse
```
Comment (ftshijt) [inline egs2/msp_podcast/ser1/local/score_ser.py:1]: If it is general, you may consider putting it into pyscripts

File: egs2/msp_podcast/ser1/run.sh [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+#   Remove the utterances with the specified emotional labels.
+#   If specifying two or more labels, concatenate them with "_" (e.g. xxx_exc_fru_fea_sur).
+#   emotional labels: ang (anger), hap (happiness), exc (excitement), sad (sadness),
+#   fru (frustration), fea (fear), sur (surprise), neu (neutral), and xxx (other)
+#--convert_to_sentiment
+#   Convert emotion to sentiment (Positive, Negative and Neutral)
+#   mapping from emotion to sentiment is as follows:
+#   Positive: hap, exc, sur
+#   Negative: ang, sad, fru, fea
+#   Neutral: neu
```
Comment (ftshijt) [inline egs2/msp_podcast/ser1/run.sh:32]: Given the different definition for each dataset, I guess you might need to have an option to control the categories of emotion?

File: egs2/msp_podcast/ser1/ser.sh [hunk 8]
```diff
@@ -0,0 +1,1677 @@
+#!/usr/bin/env bash
```
Comment (ftshijt) [inline egs2/msp_podcast/ser1/ser.sh:1]: Instead of copying the ser, please use symlink

File: espnet2/bin/ser_inference.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+        device: str = "cpu",
+        batch_size: int = 1,
+        dtype: str = "float32",
+        beam_size: int = 20,
+    ):
+
+        task = SERTask
+
+        # 1. Build SER model
+        scorers = {}
```
Comment (ftshijt) [inline espnet2/bin/ser_inference.py:66]: Is beamsize necessary for the application (also for scorers)

File: espnet2/ser/pooling/abs_pooling.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+from abc import ABC, abstractmethod
```
Comment (ftshijt) [inline espnet2/ser/pooling/abs_pooling.py:1]: It is okay to reuse existing modules (if no significant changes require for the design)

**Review by ftshijt — COMMENTED** [review]
Many thanks for the great contribution. I left some comments as follows (mostly related to redundant components) -> if we choose to expand to another task from existing ones, one would be especially beneficial if we can make some reduction to the heavy process~

**Comment by sw005320** [comment]
@Subohao, can you reflect @ftshijt's comments?

File: egs2/TEMPLATE/ser1/ser.sh (hunk 1 above)
Comment (Subohao) [reply egs2/TEMPLATE/ser1/ser.sh]: Since now it's temporary for MSP_Podcast, it's only used for these emotion categories.

File: espnet2/ser/pooling/abs_pooling.py (hunk 10 above)
Comment (Subohao) [reply espnet2/ser/pooling/abs_pooling.py]: Just create a symlink to the spk pooling function insteadly.

File: espnet2/bin/ser_inference.py (hunk 9 above)
Comment (Subohao) [reply espnet2/bin/ser_inference.py]: Remove the argument.

File: egs2/msp_podcast/ser1/ser.sh (hunk 8 above)
Comment (Subohao) [reply egs2/msp_podcast/ser1/ser.sh]: Done

File: egs2/msp_podcast/ser1/local/score_ser.py (hunk 6 above)
Comment (Subohao) [reply egs2/msp_podcast/ser1/local/score_ser.py:1]: Moved

**Comment by sw005320** [comment]
Please fix the errors
https://github.com/espnet/espnet/actions/runs/15715129680/job/44282859093?pr=6096

**Comment by sw005320** [comment]
Can you add the unit test and integration tests?
You can find some examples by checking a new task PR.

**Comment by sw005320** [comment]
The test took too long.
https://github.com/espnet/espnet/actions/runs/15882286538/job/44785536230?pr=6096#step:9:3174

Please do either of them
1. Reduce the time by using a smaller model or shorter file (recommended)
2. Increase the test time.

File: egs2/mini_an4/ser1/conf/tuning/train_asr_wavlm.yaml
Comment (sw005320) [inline egs2/mini_an4/ser1/conf/tuning/train_asr_wavlm.yaml:1]: I'm concerned that this config would be too much as an integration test.
If it were a problem in the future, I'd request that you remove the `wavlm` part

File: espnet2/ser/loss/cross_entropy_loss.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+from espnet2.ser.loss.abs_loss import AbsLoss
+
+
+class Xnt(AbsLoss):
+    """Additive angular margin softmax.
+
+    Paper: Deng, Jiankang, et al. "Arcface: Additive angular margin loss for
+    deep face recognition." Proceedings of the IEEE/CVF conference on computer
+    vision and pattern recognition. 2019.
```
Comment (sw005320) [inline espnet2/ser/loss/cross_entropy_loss.py]: Is it an angular margin softmax?
It looks like a normal softmax function.

File: espnet2/ser/pooling/mean_pooling.py
Comment (sw005320) [inline espnet2/ser/pooling/mean_pooling.py:1]: Did you care about the padded region?

File: espnet2/tasks/ser.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+from espnet2.asr.preencoder.linear import LinearProjection
+from espnet2.asr.preencoder.sinc import LightweightSincConvs
+from espnet2.asr.specaug.abs_specaug import AbsSpecAug
+from espnet2.asr.specaug.specaug import SpecAug
+from espnet2.ser.espnet_model import ESPnetSERModel
+from espnet2.ser.loss.cross_entropy_loss import Xnt
+from espnet2.ser.pooling.abs_pooling import AbsPooling
+from espnet2.ser.pooling.mean_pooling import MeanPooling
+from espnet2.ser.projector.abs_projector import AbsProjector
+from espnet2.ser.projector.linear_projector import LinearProjector
```
Comment (sw005320) [inline espnet2/tasks/ser.py:45]: Many functions are similar to those in SID.
Please consider reusing them.

File: espnet2/train/preprocessor.py [hunk 13]
```diff
@@ -638,6 +638,64 @@ def _text_process(
         return data
 
 
+class SERPreprocessor(CommonPreprocessor):
```
Comment (sw005320) [inline espnet2/train/preprocessor.py:641]: Do you need a SER-specific preprocessor?
If so, can you explain it and add it to the comment?

**Comment by sw005320** [comment]
Please work on this PR.

**Comment by sw005320** [comment]
@Subohao, please work on this PR

File: espnet2/ser/loss/cross_entropy_loss.py (hunk 11 above)
Comment (Subohao) [reply espnet2/ser/loss/cross_entropy_loss.py]: Just remove the incorrect description in the function.

File: espnet2/ser/pooling/mean_pooling.py
Comment (Subohao) [reply espnet2/ser/pooling/mean_pooling.py:1]: Add a part of taking care of the padded region.

File: espnet2/tasks/ser.py (hunk 12 above)
Comment (Subohao) [reply espnet2/tasks/ser.py:45]: I create a symbolic link to the SID if there exists same functions.

File: espnet2/train/preprocessor.py (hunk 13 above)
Comment (Subohao) [reply espnet2/train/preprocessor.py:641]: Create a commit to include the reason for SER-specific preprocessor.

**Comment by sw005320** [comment]
@Subohao, please reflect the code reviews' comments

**Comment by Subohao** [comment]
@sw005320 Would you mind checking the PR  and merge it when you are available? After modifying the errors and reflecting the comments, the PR finally passed all the CI test. Thank you!
